File: 21.merge-two-sorted-lists.c
/*
 * @lc app=leetcode id=21 lang=c
 *
 * [21] Merge Two Sorted Lists
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2){
    if(l1 == NULL)
        return l2;
    else if(l2 == NULL)
        return l1;
    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    if(l1->val < l2->val){
        dummy->next = l1;
        l1 = l1->next;
    }else{
        dummy->next = l2;
        l2 = l2->next;
    }
    struct ListNode* cur = dummy->next;
    while(l1 || l2){
        if(l1 == NULL){
            cur->next = l2;
            break;
        }else if(l2 == NULL){
            cur->next = l1;
            break;
        }
        if(l1->val < l2->val){
            cur->next = l1;
            l1 = l1->next;
        }else{
            cur->next = l2;
            l2 = l2->next;
        }
        cur = cur->next;
    }
    return dummy->next;
}

File: 21.merge-two-sorted-lists.py
#
# @lc app=leetcode id=21 lang=python3
#
# [21] Merge Two Sorted Lists
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode(0)
        dummy = head
        if l1 == None:
            return l2
        elif l2 == None:
            return l1
        while True:
            if l1 == None:
                dummy.next = l2
                break
            elif l2 == None:
                dummy.next = l1
                break
            if l1.val < l2.val:
                dummy.next = l1
                l1 = l1.next
            else:
                dummy.next = l2
                l2 = l2.next
            dummy = dummy.next
        return head.next
            

